virttest/utils_nbd.py

```diff
@@ -0,0 +1,202 @@
+"""
+Helpers for export one image by qemu-nbd,which can be consumed by VM.
+qemu-img create -f qcow2 /tmp/disk.qcow2 1G
+disk xml:
+    <disk type='network' device='disk'>
+      <driver name='qemu' type='qcow2'/>
+      <source protocol='nbd' tls='no'>
+        <host name='localhost' port='10001'/>
+      </source>
+      <target dev='sda' bus='scsi'/>
+    </disk>
+"""
+import logging
+import os
+import shutil
+
+from avocado.utils import process
+
+from virttest import utils_config
+from virttest import utils_libvirtd
+
+from virttest.utils_conn import build_server_key, build_CA, build_client_key
+
+
+class NbdExport(object):
+    """qemu-nbd export disk images.
+
+    use it by referring to below steps:
+    1) nbd = NbdExport("/var/lib/libvirt/images/nbd.qcow2", port=nbd_server_port, export_name=export_name, tls=tls_enabled)
+    2) nbd.start_nbd_server()
+       ....
+    3) nbd.cleanup()
+    """
+
+    def __init__(self, image, image_format="raw", image_size="1G", port="10001", export_name=None, tls=False, deleteExisted=True):
+        """Create a new NbdExport instance
+
+        :param image: instance of image created
+        :param image_format: image format,e.g raw or qcom2
+        :param image_size: image size
+        :param port: port where qemu-nbd export image
+        :param export_name: exported alias name
+        :param tls: whether use tls for encryption communication
+        :param deleteExisted: whether delete currently existed image file
+        """
+        # Input
+        self.image = image
+        self.image_format = image_format
+        self.image_size = image_size
+        self.port = port
+        self.export_name = export_name
+        self.tls = tls
+        if self.tls:
+            self.qemu_conf = None
+            self.libvirtd = None
+        self.deleteExisted = deleteExisted
+
+    def _create_img(self):
+        """Create a image file with specified format"""
+        if os.path.exists(self.image):
+            logging.info('image already existed')
+            if self.deleteExisted:
+                os.remove(self.image)
+            else:
+                return
+        logging.debug("create one image .... if not existed")
+        process.run("qemu-img create" + ' -f %s %s %s' % (self.image_format, self.image, self.image_size),
+                    ignore_status=True, shell=True, verbose=True)
+
+    def update_nbd_tls_config(self):
+        """update nbd tls config in qemu.conf"""
+        self.qemu_conf = utils_config.LibvirtQemuConfig()
+        client_cert_dir = "/etc/pki/libvirt-nbd"
+        self.qemu_conf.nbd_tls_x509_cert_dir = "%s" % client_cert_dir
+        self.qemu_conf.nbd_tls = True
+        self.libvirtd = utils_libvirtd.Libvirtd()
+        self.libvirtd.restart()
```

Comment:
Since libvirt 5.6, there is a  socket named libvirtd-tls.socket which controls tls connections. 
Is it necessary to restart libvirtd-tls.socket before libvirtd?

Reply:
this is generic issue we need address, leave it to common solution

Reply:
Dose that mean scripts with this PR will be failed on libvirt >= 5.6 for now?

Reply:
nbd is supported on relatively early libvirt version. Moreover, whether you support or not depends on the case in tp-libvirt.